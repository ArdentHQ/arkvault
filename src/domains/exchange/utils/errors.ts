interface ErrorWithResponse extends Error {
	response: any;
}

const isInvalidAddressError = (error: ErrorWithResponse) => {
	try {
		return error.response.json().error.message === "Invalid Address";
	} catch {
		return false;
	}
};

const isInvalidRefundAddressError = (error: ErrorWithResponse) => {
	try {
		return error.response.json().error.message === "Invalid Refund Address";
	} catch {
		return false;
	}
};

const isUnavailablePairError = (error: ErrorWithResponse) => {
	try {
		return error.response.json().error.message === "Unavailable Pair";
	} catch {
		return false;
	}
};

export { isInvalidAddressError, isInvalidRefundAddressError, isUnavailablePairError };
