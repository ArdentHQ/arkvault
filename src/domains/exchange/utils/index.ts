export * from "./assertions";
export * from "./errors";
