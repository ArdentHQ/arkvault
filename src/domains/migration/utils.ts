import { MigrationTransactionStatus } from "@/domains/migration/migration.contracts";

export const getIcon = (status: MigrationTransactionStatus) => {
	if (status === MigrationTransactionStatus.Confirmed) {
		return {
			color: "text-theme-success-600",
			name: "CircleCheckMark",
		};
	}

	return {
		color: "text-theme-warning-300",
		name: "Clock",
	};
};
