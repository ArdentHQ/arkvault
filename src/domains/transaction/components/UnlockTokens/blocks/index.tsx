export { UnlockTokensAuthentication } from "./UnlockTokensAuthentication";
export { UnlockTokensReview } from "./UnlockTokensReview";
export { UnlockTokensSelect } from "./UnlockTokensSelect";
export { UnlockTokensSummary } from "./UnlockTokensSummary";
