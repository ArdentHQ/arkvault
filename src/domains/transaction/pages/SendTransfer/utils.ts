export const getFeeType = (numberOfRecipients: number): "multiPayment" | "transfer" =>
	numberOfRecipients > 1 ? "multiPayment" : "transfer";
