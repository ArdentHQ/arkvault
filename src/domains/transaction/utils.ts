import { Services } from "@ardenthq/sdk";
import { RecipientItem } from "@/domains/transaction/components/RecipientList/RecipientList.contracts";

export const isNoDeviceError = (error: any) => {
	if (!error) {
		return false;
	}

	if (String(error).includes("no device found")) {
		return true;
	}

	return false;
};

export const isRejectionError = (error: any) => String(error).includes("Condition of use not satisfied");

export const handleBroadcastError = ({ errors }: Services.BroadcastResponse) => {
	const allErrors = Object.values(errors);

	if (allErrors.length === 0) {
		return;
	}

	throw new Error(allErrors[0]);
};

export const getTransferType = ({ recipients }: { recipients: RecipientItem[] }): "multiPayment" | "transfer" => {
	const isMultiPayment = recipients.length > 1;
	return isMultiPayment ? "multiPayment" : "transfer";
};

export const withAbortPromise =
	(signal?: AbortSignal, callback?: () => void) =>
	<T>(promise: Promise<T>) =>
		new Promise<T>((resolve, reject) => {
			if (signal) {
				signal.addEventListener("abort", () => {
					callback?.();
					reject("ERR_ABORT");
				});
			}

			return promise.then(resolve).catch(reject);
		});
