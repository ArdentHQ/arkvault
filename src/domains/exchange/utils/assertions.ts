import { AssertionError } from "assert";
import { Contracts } from "@ardenthq/sdk-profiles";

import { CurrencyData } from "@/domains/exchange/exchange.contracts";
import { ExchangeService } from "@/domains/exchange/services/exchange.service";

export function assertExchangeService(exchangeService?: ExchangeService): asserts exchangeService is ExchangeService {
	if (!(exchangeService instanceof ExchangeService)) {
		throw new AssertionError({
			message: `Expected 'exchangeService' to be ExchangeService, but received ${exchangeService}`,
		});
	}
}

export function assertExchangeTransaction(
	exchangeTransaction?: Contracts.IExchangeTransaction,
): asserts exchangeTransaction is Contracts.IExchangeTransaction {
	if (typeof exchangeTransaction?.orderId !== "function") {
		throw new AssertionError({
			message: `Expected 'exchangeTransaction' to be Contracts.IExchangeTransaction, but received ${exchangeTransaction}`,
		});
	}
}

export function assertCurrency(currencyData?: CurrencyData): asserts currencyData is CurrencyData {
	if (!currencyData || typeof currencyData !== "object" || !("coin" in currencyData && "name" in currencyData)) {
		throw new AssertionError({
			message: `Expected 'currencyData' to be CurrencyData, but received ${currencyData}`,
		});
	}
}
