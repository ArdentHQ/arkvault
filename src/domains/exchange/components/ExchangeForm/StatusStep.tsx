import { Contracts } from "@ardenthq/sdk-profiles";
import React, { useEffect } from "react";
import { Trans, useTranslation } from "react-i18next";

import { ExchangeStatus } from "./StatusStep.blocks";
import { Amount } from "@/app/components/Amount";
import { Clipboard } from "@/app/components/Clipboard";
import { Icon } from "@/app/components/Icon";
import { Link } from "@/app/components/Link";
import { TruncateMiddleDynamic } from "@/app/components/TruncateMiddleDynamic";
import { useExchangeContext } from "@/domains/exchange/contexts/Exchange";
import { useOrderStatus } from "@/domains/exchange/hooks/use-order-status";
import { delay } from "@/utils/delay";

interface StatusStepProperties {
	transferTransactionId: string|undefined;
	exchangeTransaction: Contracts.IExchangeTransaction;
	onUpdate: (orderId: string, parameters: any) => void;
}

export const StatusStep = ({ exchangeTransaction, onUpdate, transferTransactionId }: StatusStepProperties) => {
	const { t } = useTranslation();

	const { provider: exchangeProvider } = useExchangeContext();

	const { checkOrderStatus, prepareParameters } = useOrderStatus();

	useEffect(() => {
		let timeout: NodeJS.Timeout;

		const fetchStatus = async () => {
			const responses = await checkOrderStatus([exchangeTransaction]);

			/* istanbul ignore else -- @preserve */
			if (responses) {
				const orderStatus = responses[exchangeTransaction.orderId()];

				// status changed
				if (orderStatus.status !== exchangeTransaction.status()) {
					const parameters = prepareParameters(exchangeTransaction, orderStatus);
					onUpdate(exchangeTransaction.id(), parameters);
				}

				// fetch again if pending
				if (orderStatus.status < Contracts.ExchangeTransactionStatus.Finished) {
					timeout = delay(fetchStatus, 15_000);
				}
			}
		};

		fetchStatus();

		return () => clearTimeout(timeout);
	}, [checkOrderStatus, exchangeTransaction, onUpdate, prepareParameters]);

	return (
		<div data-testid="ExchangeForm__status-step" className="flex flex-col">
			<div className="flex items-center space-x-1">
				<span className="text-xs font-semibold text-theme-secondary-500 dark:text-theme-secondary-700">
					{t("EXCHANGE.TRANSACTION_ID")}: {exchangeTransaction.orderId()}
				</span>
				<span className="flex text-theme-primary-300 dark:text-theme-secondary-600">
					<Clipboard variant="icon" data={exchangeTransaction.orderId()}>
						<Icon name="Copy" />
					</Clipboard>
				</span>
			</div>

			<div className="mt-3 flex flex-col rounded-xl border border-theme-secondary-300 px-6 py-5 dark:border-theme-secondary-800">
				<div className="flex flex-col space-y-2">
					<span className="text-sm font-semibold text-theme-secondary-500 dark:text-theme-secondary-700">
						{t("EXCHANGE.EXCHANGE_FORM.YOU_SEND")}
					</span>
					<Amount
						value={exchangeTransaction.input().amount}
						ticker={exchangeTransaction.input().ticker}
						className="text-lg font-semibold"
					/>
				</div>

				<div className="mt-4 flex flex-col space-y-2 border-t border-theme-secondary-300 pt-4 dark:border-theme-secondary-800">
					<span className="text-sm font-semibold text-theme-secondary-500 dark:text-theme-secondary-700">
						{t("EXCHANGE.TO_ADDRESS")}
					</span>
					<div className="flex items-center space-x-2 whitespace-nowrap text-lg font-semibold">
						<TruncateMiddleDynamic value={exchangeTransaction.input().address} className="no-ligatures" />
						<span className="flex text-theme-primary-300 dark:text-theme-secondary-600">
							<Clipboard variant="icon" data={exchangeTransaction.input().address}>
								<Icon name="Copy" />
							</Clipboard>
						</span>
					</div>
				</div>

				{transferTransactionId &&
					<div
						className="mt-4 flex flex-col space-y-2 border-t border-theme-secondary-300 pt-4 dark:border-theme-secondary-800">
						<span className="text-sm font-semibold text-theme-secondary-500 dark:text-theme-secondary-700">
							{t("EXCHANGE.ARK_TRANSACTION_ID")}
						</span>
						<div className="flex items-center space-x-2 whitespace-nowrap text-lg font-semibold">
							<TruncateMiddleDynamic value={transferTransactionId} className="no-ligatures"/>
							<span className="flex text-theme-primary-300 dark:text-theme-secondary-600">
								<Clipboard variant="icon" data={transferTransactionId}>
									<Icon name="Copy"/>
								</Clipboard>
							</span>
						</div>
					</div>}
			</div>

			<ExchangeStatus exchangeTransaction={exchangeTransaction}/>

			<div
				className="-mx-10 border-t border-dashed border-theme-secondary-300 px-10 pt-6 dark:border-theme-secondary-800"/>

			<div className="flex flex-col space-y-4">
				<div className="flex flex-col space-y-2">
					<span className="text-sm font-semibold text-theme-secondary-500 dark:text-theme-secondary-700">
						{t("EXCHANGE.EXCHANGE_FORM.YOU_GET")}
					</span>
					<span>
						â‰ˆ{" "}
						<Amount
							value={exchangeTransaction.output().amount}
							ticker={exchangeTransaction.output().ticker}
							className="font-semibold"
						/>
					</span>
				</div>

				<div className="flex flex-col space-y-2">
					<span className="text-sm font-semibold text-theme-secondary-500 dark:text-theme-secondary-700">
						{t("EXCHANGE.TO_ADDRESS")}
					</span>
					<div className="flex items-center space-x-2 whitespace-nowrap font-semibold">
						<TruncateMiddleDynamic value={exchangeTransaction.output().address} className="no-ligatures" />
						<span className="flex text-theme-primary-300 dark:text-theme-secondary-600">
							<Clipboard variant="icon" data={exchangeTransaction.output().address}>
								<Icon name="Copy" />
							</Clipboard>
						</span>
					</div>
				</div>
			</div>

			<div className="mt-6 rounded-lg bg-theme-secondary-100 px-6 py-3 text-xs dark:bg-theme-secondary-800">
				<Trans
					i18nKey="EXCHANGE.EXCHANGE_FORM.SUPPORT_INFO"
					values={{
						email: exchangeProvider?.emailAddress,
						exchange: exchangeProvider?.name,
					}}
					components={{ linkEmail: <Link to={`mailto:${exchangeProvider?.emailAddress}`} isExternal /> }}
				/>
			</div>
		</div>
	);
};
