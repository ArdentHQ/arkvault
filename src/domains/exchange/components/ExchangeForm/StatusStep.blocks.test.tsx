import { Contracts } from "@payvo/sdk-profiles";
import { renderHook } from "@testing-library/react-hooks";
import React from "react";
import { useTranslation } from "react-i18next";

import { ExchangeStatus } from "./StatusStep.blocks";
import { env, getDefaultProfileId, render, screen } from "@/utils/testing-library";

let profile: Contracts.IProfile;
let exchangeTransaction: Contracts.IExchangeTransaction;

describe("ExchangeStatus", () => {
	beforeAll(() => {
		profile = env.profiles().findById(getDefaultProfileId());
		exchangeTransaction = profile.exchangeTransactions().create({
			input: {
				address: "inputAddress",
				amount: 1,
				ticker: "btc",
			},
			orderId: "orderId",
			output: {
				address: "outputAddress",
				amount: 100,
				ticker: "ark",
			},
			provider: "provider",
		});
	});

	it("should show failed status", () => {
		const { result } = renderHook(() => useTranslation());
		const { t } = result.current;

		exchangeTransaction.setStatus(Contracts.ExchangeTransactionStatus.Failed);
		const { container } = render(<ExchangeStatus exchangeTransaction={exchangeTransaction} />);

		expect(container).toHaveTextContent("circle-cross.svg");
		expect(container).toHaveTextContent(t("EXCHANGE.TRANSACTION_STATUS.FAILED"));

		expect(container).toMatchSnapshot();
	});

	it("should show refunded status", () => {
		const { result } = renderHook(() => useTranslation());
		const { t } = result.current;

		exchangeTransaction.setStatus(Contracts.ExchangeTransactionStatus.Refunded);
		const { container } = render(<ExchangeStatus exchangeTransaction={exchangeTransaction} />);

		expect(container).toHaveTextContent("circle-exclamation-mark.svg");
		expect(container).toHaveTextContent(t("EXCHANGE.TRANSACTION_STATUS.REFUNDED"));

		expect(container).toMatchSnapshot();
	});

	it("should show verifying status", () => {
		const { result } = renderHook(() => useTranslation());
		const { t } = result.current;

		exchangeTransaction.setStatus(Contracts.ExchangeTransactionStatus.Verifying);
		const { container } = render(<ExchangeStatus exchangeTransaction={exchangeTransaction} />);

		expect(container).toHaveTextContent("circle-exclamation-mark.svg");
		expect(container).toHaveTextContent(t("EXCHANGE.TRANSACTION_STATUS.VERIFYING"));

		expect(container).toMatchSnapshot();
	});

	it("should show expired status", () => {
		const { result } = renderHook(() => useTranslation());
		const { t } = result.current;

		exchangeTransaction.setStatus(Contracts.ExchangeTransactionStatus.Expired);
		const { container } = render(<ExchangeStatus exchangeTransaction={exchangeTransaction} />);

		expect(container).toHaveTextContent("circle-cross.svg");
		expect(container).toHaveTextContent(t("EXCHANGE.TRANSACTION_STATUS.EXPIRED"));

		expect(container).toMatchSnapshot();
	});

	it("should show exchange status", () => {
		exchangeTransaction.setStatus(Contracts.ExchangeTransactionStatus.Confirming);
		const { rerender } = render(<ExchangeStatus exchangeTransaction={exchangeTransaction} />);

		expect(screen.queryByTestId("StatusIcon__check-mark")).not.toBeInTheDocument();
		expect(screen.getAllByTestId("StatusIcon__spinner")).toHaveLength(1);
		expect(screen.getAllByTestId("StatusIcon__empty")).toHaveLength(2);

		exchangeTransaction.setStatus(Contracts.ExchangeTransactionStatus.Exchanging);
		rerender(<ExchangeStatus exchangeTransaction={exchangeTransaction} />);

		expect(screen.getAllByTestId("StatusIcon__check-mark")).toHaveLength(1);
		expect(screen.getAllByTestId("StatusIcon__spinner")).toHaveLength(1);
		expect(screen.getAllByTestId("StatusIcon__empty")).toHaveLength(1);

		exchangeTransaction.setStatus(Contracts.ExchangeTransactionStatus.Sending);
		rerender(<ExchangeStatus exchangeTransaction={exchangeTransaction} />);

		expect(screen.getAllByTestId("StatusIcon__check-mark")).toHaveLength(2);
		expect(screen.getAllByTestId("StatusIcon__spinner")).toHaveLength(1);
		expect(screen.queryByTestId("StatusIcon__empty")).not.toBeInTheDocument();

		exchangeTransaction.setStatus(Contracts.ExchangeTransactionStatus.Finished);
		rerender(<ExchangeStatus exchangeTransaction={exchangeTransaction} />);

		expect(screen.getAllByTestId("StatusIcon__check-mark")).toHaveLength(3);
		expect(screen.queryByTestId("StatusIcon__spinner")).not.toBeInTheDocument();
		expect(screen.queryByTestId("StatusIcon__empty")).not.toBeInTheDocument();
	});
});
