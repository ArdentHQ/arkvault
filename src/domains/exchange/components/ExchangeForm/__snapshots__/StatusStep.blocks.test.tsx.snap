// Vitest Snapshot v1

exports[`ExchangeStatus > should show expired status 1`] = `
<div>
  <div
    class="my-6 css-mpc03v"
  >
    <div
      class="css-1f7sa7y"
    >
      <div
        class="css-v0ob3f"
        height="16"
        width="16"
      >
        <svg
          id="circle-cross"
          viewBox="0 0 20 20"
          x="0"
          xml:space="preserve"
          xmlns="http://www.w3.org/2000/svg"
          y="0"
        >
          <g
            fill="none"
            stroke="currentColor"
            stroke-width="2"
          >
            <circle
              cx="10"
              cy="10"
              r="9"
              stroke-miterlimit="10"
            />
            <path
              d="M7 13l6-6M13 13L7 7"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </g>
        </svg>
      </div>
      <span>
        Error
      </span>
    </div>
    <div
      class="css-1h3dpsb"
    >
      This transaction has expired.
    </div>
  </div>
</div>
`;

exports[`ExchangeStatus > should show failed status 1`] = `
<div>
  <div
    class="my-6 css-mpc03v"
  >
    <div
      class="css-1f7sa7y"
    >
      <div
        class="css-v0ob3f"
        height="16"
        width="16"
      >
        <svg
          id="circle-cross"
          viewBox="0 0 20 20"
          x="0"
          xml:space="preserve"
          xmlns="http://www.w3.org/2000/svg"
          y="0"
        >
          <g
            fill="none"
            stroke="currentColor"
            stroke-width="2"
          >
            <circle
              cx="10"
              cy="10"
              r="9"
              stroke-miterlimit="10"
            />
            <path
              d="M7 13l6-6M13 13L7 7"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </g>
        </svg>
      </div>
      <span>
        Error
      </span>
    </div>
    <div
      class="css-1h3dpsb"
    >
      This transaction has failed.
    </div>
  </div>
</div>
`;

exports[`ExchangeStatus > should show refunded status 1`] = `
<div>
  <div
    class="my-6 css-mpc03v"
  >
    <div
      class="css-1hv657z"
    >
      <div
        class="css-v0ob3f"
        height="16"
        width="16"
      >
        <svg
          id="circle-exclamation-mark"
          viewBox="0 0 20 20"
          x="0"
          xml:space="preserve"
          xmlns="http://www.w3.org/2000/svg"
          y="0"
        >
          <g
            fill="none"
            stroke="currentColor"
            stroke-width="2"
          >
            <circle
              cx="10"
              cy="10"
              r="9"
              stroke-miterlimit="10"
            />
            <path
              d="M10 10.5v-5"
              stroke-linecap="round"
              stroke-miterlimit="10"
            />
            <path
              d="M10 14h0"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </g>
        </svg>
      </div>
      <span>
        Warning
      </span>
    </div>
    <div
      class="css-rfibfr"
    >
      This transaction has been refunded.
    </div>
  </div>
</div>
`;

exports[`ExchangeStatus > should show verifying status 1`] = `
<div>
  <div
    class="my-6 css-mpc03v"
  >
    <div
      class="css-1hv657z"
    >
      <div
        class="css-v0ob3f"
        height="16"
        width="16"
      >
        <svg
          id="circle-exclamation-mark"
          viewBox="0 0 20 20"
          x="0"
          xml:space="preserve"
          xmlns="http://www.w3.org/2000/svg"
          y="0"
        >
          <g
            fill="none"
            stroke="currentColor"
            stroke-width="2"
          >
            <circle
              cx="10"
              cy="10"
              r="9"
              stroke-miterlimit="10"
            />
            <path
              d="M10 10.5v-5"
              stroke-linecap="round"
              stroke-miterlimit="10"
            />
            <path
              d="M10 14h0"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </g>
        </svg>
      </div>
      <span>
        Warning
      </span>
    </div>
    <div
      class="css-rfibfr"
    >
      This transaction is being verified by the provider.
    </div>
  </div>
</div>
`;
